onnxruntime/core/providers/cuda/reduction/reduction_functions.cu

```diff
@@ -190,107 +246,122 @@ __global__ void reduce_all_kernel(const int size, const TIn * data, TOut* output
       // Compilation time if-else branch controlled by template argument can be
       // optimized out, so there will be no branch in real computation phase.
       if (DivideResultBySize) {
-        output[0] = TFinalOp()(buffer[0] / TOut(size));
+        output[0] = TFinalOp()(TOut(block_reductions_buffer[0]) / TOut(row_size));
       } else {
-        output[0] = TFinalOp()(buffer[0]);
+        output[0] = TFinalOp()(TOut(block_reductions_buffer[0]));
       }
     }
   }
 }
 
-template<typename TIn, typename TOut, typename TOp, typename TFinalOp, bool DivideResultBySize>
-void call_reduce_all_kernel(const TIn *data, TOut *output, int size, TOut *buffer) {
-  const auto block_size = compute_block_size(size);
-  const int num_blocks = compute_grid_size(size);
-  const dim3 block(block_size.first, block_size.second, 1);
-  const dim3 grid(num_blocks, 1, 1);
+template <typename TIn, typename TOut, typename TBuf, typename TOp, typename TFinalOp, bool DivideResultBySize>
+__global__ void reduce_all_in_rows_kernel(
+    const int num_rows, const int row_size, const TIn* const input, TOut* const output,
+    TBuf* const block_reductions_buffer, int* const block_done_counts_buffer) {
+  const int num_blocks_in_grid_row = gridDim.x;
+  const int row_id_in_grid = blockIdx.y;
+  const int num_grid_rows = gridDim.y;
+
+  // one row per iteration
+  // row_id is int64_t to avoid int overflow in offset calculations
+  for (int64_t row_id = row_id_in_grid; row_id < num_rows; row_id += num_grid_rows) {
+    const TIn* const row_data = input + row_id * row_size;
+    TOut* const row_output = output + row_id;
+    TBuf* const row_block_reductions_buffer = block_reductions_buffer + row_id * num_blocks_in_grid_row;
+    int* const row_block_done_counts_buffer = block_done_counts_buffer + row_id;
+
+    reduce_all_in_row<TIn, TOut, TBuf, TOp, TFinalOp, DivideResultBySize>(
+        row_size, row_data, row_output,
+        row_block_reductions_buffer, row_block_done_counts_buffer);
+  }
+}
+
+template <typename TIn, typename TOut, typename TOp, typename TFinalOp, bool DivideResultBySize>
+Status call_reduce_all_in_rows_kernel(
```

Comment:
>call_reduce_all_in_rows_kernel [](start = 7, length = 30)

Please rename it to `call_reduce_matrix_columns`

Reply:
good idea. i updated them, see if the new ones are ok

---
In reply to: [517599158](https://github.com/microsoft/onnxruntime/pull/5639#discussion_r517599158) [](ancestors = 517599158)